game/src/plugins/skills/mining/src/gem.kt

```diff
@@ -1,13 +1,18 @@
 package org.apollo.game.plugin.skills.mining
 
-data class Gem(val id: Int, val chance: Int)
-
-private val GEMS = mutableMapOf<Int, Gem>()
+enum class Gem(val id: Int, val chance: Int) {
+    UNCUT_SAPPHIRE(1623, 0), // uncut sapphire
+    UNCUT_EMERALD(1605, 0),// uncut emerald
+    UNCUT_RUBY(1619, 0), // uncut ruby
+    UNCUT_DIAMOND(1617, 0)  // uncut diamond
+}
 
 fun lookupGem(id: Int): Gem? {
-    return GEMS[id]
+    for (gem in Gem.values()) {
```

Comment:
Try and use kotlin features when writing in kotlin, like replacing this entire function with:

```kotlin
fun lookupGem(id: Int): Gem? = Gem.values().find { it.id == id }
```

by combining a one-line function definition with kotlin's functional stuff

Reply:
Thanks. I did not know about this. I am still trying to get used to all the one liners in Kotlin.

Reply:
@Major- Wouldn't Gem.values() create a new array every time `lookupGem` is called then?

Reply:
Yes.  Arrays of enum values should be cached.

Reply:
So does a global array need to be created to store the array from the enum? What would be the best way to cache it? If we still have to populate an array when using an enum, what advantage is gained by using them over a data class? We will now have two copies of the data instead of one.

I would also like to note that an enum generates a class at compile time so both and enum and data class are stored as objects and will take the same amount of space in memory when storing the same data.

Reply:
I don't personally know Kotlin, but in Java you typically do `public static final Gem VALUES = Gem.values()`.

Reply:
So for the sake of getting this done with I just created a val and stored the array.

Reply:
amulet of glory with charges should effect chances of mining gems.

Reply:
There are several things that will. But at the moment we are just working on base functionality.